client/adventureMap/AdventureMapInterface.cpp

```diff
@@ -159,6 +155,8 @@ void AdventureMapInterface::deactivate()
 
 void AdventureMapInterface::showAll(Canvas & to)
 {
+	refreshDisplayedPath();
+
```

Comment:
Might be better to place it in required netpack appliers / CPlayerInterface.
showAll() should be used only to display stuff.

Reply:
Hi Ivan,

I have some doubts if this is the proper way to deal with that. After digging into the code got some observations:

- Shutting down InfoWindow updates the displayed route properly. That's why it works for level 1 dwellings.

- Interface not updated after closing CWindowObject is getting the update on closing (any) popup window.

It looks like the issue might be related to how the display is refreshed when windows are closed. In the destructor of CWindowObject, I see that show() is called to change the cursor visibility:

```
CWindowObject::~CWindowObject()
{
    if (options & RCLICK_POPUP)
        CCS->curh->show();
}
```
Inside show(), the SDL_ShowCursor() function is called, which seems to be triggering some kind of screen refresh:

```
void CursorHandler::show()
{
    if (showing)
        return;

    showing = true;
    cursor->setVisible(true);
}
```

And:

```
void CursorHardware::setVisible(bool on)
{
    GH.dispatchMainThread([on]()
    {
        if (on)
            SDL_ShowCursor(SDL_ENABLE);
        else
            SDL_ShowCursor(SDL_DISABLE);
    });
}
```

Perhaps there should be some function that calls SDL to be updated just before leaving destructors body ? Or this is due to some threads code ?

I also tried to get it done how `InfoWindow::close()` does it but with no success :(

Let me know which approach we should take to resolve this issue — the current approach or the one related to above.

Reply:
ok so... it should be fixed like here https://github.com/vcmi/vcmi/pull/3685/commits/96b18b1b8e841bd4f3697e494be68e48a5684ccb

Reply:
Why do you need to do this on infowindow closing?

When hero accepts creatures from level 1 dwelling game sends RebalanceStacks pack.
This will in turn call ApplyClientNetPackVisitor::visitRebalanceStacks on client
Which calls CPlayerInterface::garrisonsChanged

Updating path must be done in one of these methods. And `visitRebalanceStacks` does have cl.invalidatePaths()

When hero purchases creatures from a dwelling, game sends InsertNewStack pack, which also results in the same call chain including cl.invalidatePaths()

Judging from your changes, I suspect what needs to be changed is that cl.invalidatePaths should also call that verifyPath method